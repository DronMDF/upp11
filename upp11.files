LICENSE
README.md
upp11.h
Makefile
test/testupp.cpp
test/testfailures.cpp
test/testuppdetail.cpp
