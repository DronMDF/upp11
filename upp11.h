
#pragma once
#include <list>
#include <iostream>

namespace upp11 {

class TestInvoker {
public:
	virtual void invoke() = 0;
};

class TestCollection {
private:
	std::list<TestInvoker *> tests;

	static TestCollection &getInstance() {
		static TestCollection collection;
		return collection;
	};

public:
	static void addTest(class TestInvoker *test)
	{
		TestCollection &collection = getInstance();
		collection.tests.push_back(test);
	}

	static void runAllTests()
	{
		TestCollection &collection = getInstance();
		for (auto *t: collection.tests) {
			t->invoke();
		}
	}
};

template <typename T>
class TestInvokerImpl : public TestInvoker {
public:
	TestInvokerImpl() {
		TestCollection::addTest(this);
	}

private:
	void invoke() override {
		T test;
		test();
	}
};

} // end of namespace upp11

#define UP_RUN() \
	upp11::TestCollection::runAllTests()

#define UP_TEST(name) \
class Test##name { \
public: \
	void operator()(); \
}; \
static upp11::TestInvokerImpl<Test##name> test##name##invoker; \
void Test##name::operator()()

#define UP_FAIL(msg) \
	std::cout << msg << std::endl;
