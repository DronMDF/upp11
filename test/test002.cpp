#include <upp11.h>

using namespace std;

UP_SUITE_BEGIN(suiteAssertExceptionWithMessage)

UP_TEST(ShouldFailByType)
{
	UP_ASSERT_EXCEPTION(int, "hello", []{
		throw runtime_error("goodby");
	});
}

UP_TEST(ShouldFailByNoThrow)
{
	UP_ASSERT_EXCEPTION(runtime_error, "hello", []{
		// no throw
	});
}

UP_TEST(ShouldFailByNotEqualMessage)
{
	UP_ASSERT_EXCEPTION(exception, "message", []{
		throw runtime_error("another message");
	});
}

UP_TEST(ShouldFailByChildException)
{
	UP_ASSERT_EXCEPTION(overflow_error, "message", []{
		throw runtime_error("message");
	});
}

UP_SUITE_END()

UP_SUITE_BEGIN(suiteAssertException)

UP_TEST(ShouldFailByType)
{
	UP_ASSERT_EXCEPTION(int, []{
		throw runtime_error("goodby");
	});
}

UP_TEST(ShouldFailByNoThrow)
{
	UP_ASSERT_EXCEPTION(runtime_error, []{
		// no throw
	});
}

UP_SUITE_END()

UP_MAIN()
