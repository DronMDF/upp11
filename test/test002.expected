test/test002.cpp(50): expected exception runtime_error not throw
suiteAssertException/ShouldFailByNoThrow: FAIL
test/test002.cpp(43): expected exception int not throw
suiteAssertException/ShouldFailByType: FAIL
test/test002.cpp(32): expected exception overflow_error("message") not throw
suiteAssertExceptionWithMessage/ShouldFailByChildException: FAIL
test/test002.cpp(18): expected exception runtime_error("hello") not throw
suiteAssertExceptionWithMessage/ShouldFailByNoThrow: FAIL
test/test002.cpp(25): check exception exception("message") failed
	catched exception: "another message"
suiteAssertExceptionWithMessage/ShouldFailByNotEqualMessage: FAIL
test/test002.cpp(11): expected exception int is not child of std::exception
suiteAssertExceptionWithMessage/ShouldFailByType: FAIL
